hubstaff-export.rb

```diff
@@ -32,29 +33,275 @@
 require 'date'
 require 'net/http'
 require 'fileutils'
+require 'forwardable'
+require 'base64'
 require 'json'
 require 'pp'
 
-class Hash
-  def compact
-    delete_if {|k,v| v.nil? }
+# a network error occurred
+class NetworkError < StandardError; end
+
+# an error from a command
+class CommandError < StandardError; end
+
+# an error from a command
+class TokenError < StandardError; end
+
+# Simple network wrapper around Net::HTTP
+class HTTPClient
+  # @param skip_ssl_verify [Boolean]
+  def initialize(skip_ssl_verify: false)
+    @skip_ssl_verify = skip_ssl_verify
+  end
+
+  def post(url, params: {}, headers: nil)
+    uri     = URI.parse(url)
+    request = Net::HTTP::Post.new(uri.request_uri, headers)
+
+    request.set_form_data(params)
+
+    parse_response(http(uri).request(request))
+  rescue Errno::ETIMEDOUT
+    raise NetworkError, 'there was a timout'
+  end
+
+  def post_json(url, params: {}, headers: nil)
+    uri                  = url.is_a?(URI) ? url : URI.parse(url)
+    request              = Net::HTTP::Post.new(uri.request_uri, headers)
+    request.content_type = 'application/json'
+    request.body         = params.to_json
+
+    parse_response(http(uri).request(request))
+  rescue Errno::ETIMEDOUT
+    raise NetworkError, 'there was a timout'
+  end
+
+  def get(url, params: {}, headers: nil)
+    uri       = url.is_a?(URI) ? url : URI.parse(url)
+    uri.query = URI.encode_www_form(params)
+    request   = Net::HTTP::Get.new(uri.request_uri, headers)
+
+    parse_response(http(uri).request(request))
+  rescue Errno::ETIMEDOUT
+    raise NetworkError, 'there was a timout'
+  end
+
+  private
+
+  def parse_response(response)
+    case response
+    when Net::HTTPOK, Net::HTTPCreated
+      JSON.parse(response.body, symbolize_names: true)
+    when Net::HTTPNotFound
+      raise NetworkError, 'page not found'
+    when Net::HTTPUnauthorized
+      raise NetworkError, 'not authorized request'
+    when Net::HTTPBadRequest
+      raise NetworkError, 'bad request'
+    when Net::HTTPTooManyRequests
+      raise NetworkError, 'rate limit reached'
+    when Net::HTTPServiceUnavailable
+      raise NetworkError, 'timeout fetching data.'
+    else
+      raise NetworkError, "Unexpected Error: #{response}"
+    end
+  end
+
+  def http(uri)
+    http             = Net::HTTP.new(uri.host, uri.port)
+    http.verify_mode = OpenSSL::SSL::VERIFY_NONE if @skip_ssl_verify
+    http.use_ssl     = true
+    http
+  end
+end
+
+# Class to manage communicating with the auth server
+class AuthClient
```

Comment:
Let's move all classes into separate files for the sake of better readability.

Reply:
Again ease of use is the reason for the tool here.  So I really didn't want to muck it up right now with multiple files.